File: weird_captcha_gym/shared_runtime/app/mechanics/patchwork_skirmish.js
(() => {
  "use strict";

  const ID = "patchwork_skirmish";
  const DIRS = {north: [-1, 0], south: [1, 0], west: [0, -1], east: [0, 1]};
  const TIE_ORDER = ["north", "west", "south", "east"];
  let model = null;

  const esc = (value) => String(value == null ? "" : value)
    .replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;");
  const live = (owner) => Object.values(model.units).filter((unit) => unit.cells.length && (owner == null || Number(unit.owner) === owner));
  const cellsKey = (cells) => new Set(cells.map(([row, col]) => `${row},${col}`));
  const distance = (a, b) => Math.min(...a.flatMap(([ar, ac]) => b.map(([br, bc]) => Math.abs(ar - br) + Math.abs(ac - bc))));
  const adjacent = (a, b) => distance(a, b) === 1;

  function occupied(skip) {
    const result = new Set();
    Object.values(model.units).forEach((unit) => {
      if (unit.id === skip) return;
      unit.cells.forEach(([row, col]) => result.add(`${row},${col}`));
    });
    return result;
  }

  function translated(cells, direction) {
    const [dr, dc] = DIRS[direction] || [0, 0];
    return cells.map(([row, col]) => [row + dr, col + dc]);
  }

  function validTranslation(unitId, direction) {
    if (!DIRS[direction]) return null;
    const cells = translated(model.units[unitId].cells, direction);
    const blocked = occupied(unitId);
    if (cells.some(([row, col]) => row < 0 || row >= model.world.height || col < 0 || col >= model.world.width || blocked.has(`${row},${col}`))) return null;
    return cells;
  }

  function nearest(unit) {
    const owner = Number(unit.owner) === 0 ? 1 : 0;
    const choices = live(owner);
    return choices.sort((a, b) => distance(unit.cells, a.cells) - distance(unit.cells, b.cells) || String(a.id).localeCompare(String(b.id)))[0] || null;
  }

  function removeEdge(unit, power) {
    const count = Math.min(Math.max(0, Number(power)), unit.cells.length);
    if (count) unit.cells.splice(unit.cells.length - count, count);
    return count;
  }

  function enemyResponse() {
    const notes = [];
    live(1).sort((a, b) => String(a.id).localeCompare(String(b.id))).forEach((enemy) => {
      if (!enemy.cells.length) return;
      let target = nearest(enemy);
      if (!target) return;
      if (adjacent(enemy.cells, target.cells)) {
        const removed = removeEdge(target, 1);
        model.enemyActions += 1;
        notes.push(`${enemy.id} chips ${target.id} (${removed} cell)`);
        return;
      }
      for (let step = 0; step < Number(model.world.enemy_movement_points); step += 1) {
        target = nearest(enemy);
        if (!target || !enemy.cells.length) break;
        const choices = Object.keys(DIRS).map((direction) => {
          const next = validTranslation(enemy.id, direction);
          return next ? [distance(next, target.cells), direction, next] : null;
        }).filter(Boolean);
        if (!choices.length) break;
        choices.sort((a, b) => a[0] - b[0] || TIE_ORDER.indexOf(a[1]) - TIE_ORDER.indexOf(b[1]));
        enemy.cells = choices[0][2];
        model.enemyActions += 1;
        notes.push(`${enemy.id} drifts ${choices[0][1]}`);
      }
    });
    model.lastResponse = notes.length ? notes.join("; ") : "RIVALS HAVE NO LEGAL RESPONSE";
  }

  function applyEvent(event) {
    if (model.terminal) throw new Error("the skirmish is already over");
    const type = String(event.type || "");
    if (type === "move") {
      const unit = model.units[String(event.unit_id || "")];
      if (!unit || Number(unit.owner) !== 0 || !unit.cells.length) throw new Error("move must select a living player patch");
      if (Number(unit.moves_left) <= 0) throw new Error("that patch has spent its movement");
      const cells = validTranslation(unit.id, String(event.direction || ""));
      if (!cells) throw new Error("move collides with the board or another patch");
      unit.cells = cells;
      unit.moves_left -= 1;
    } else if (type === "attack") {
      const unit = model.units[String(event.unit_id || "")];
      const target = model.units[String(event.target_id || "")];
      if (!unit || !target || Number(unit.owner) !== 0 || Number(target.owner) !== 1) throw new Error("attack must select a player and rival patch");
      if (!unit.cells.length || !target.cells.length || Number(unit.attacks_left) <= 0) throw new Error("that patch cannot attack now");
      if (!adjacent(unit.cells, target.cells)) throw new Error("the rival patch is not neighboring");
      removeEdge(target, model.world.attack_power);
      unit.attacks_left -= 1;
    } else if (type === "end_turn") {
      if (!live(1).length) {
        model.terminal = true;
        model.won = Boolean(live(0).length);
        return;
      }
      enemyResponse();
      model.turn += 1;
      live().filter((unit) => Number(unit.owner) === 0).forEach((unit) => {
        unit.moves_left = Number(model.world.movement_points);
        unit.attacks_left = 1;
      });
      if (!live(0).length || model.turn >= Number(model.world.max_turns)) {
        model.terminal = true;
        model.won = false;
      }
    } else {
      throw new Error("unknown patchwork action");
    }
    model.actionCount += 1;
  }

  function snapshot() {
    const units = {};
    Object.keys(model.units).sort().forEach((unitId) => {
      const unit = model.units[unitId];
      units[unitId] = {cells: unit.cells.map((cell) => [...cell]), moves_left: Number(unit.moves_left), attacks_left: Number(unit.attacks_left)};
    });
    return {turn: Number(model.turn), units, terminal: Boolean(model.terminal), won: Boolean(model.won), action_count: Number(model.actionCount), enemy_actions: Number(model.enemyActions), last_response: String(model.lastResponse)};
  }

  function event(type, details) {
    const item = {sequence: model.events.length + 1, type, input_source: type === "end_turn" ? "turn_button" : model.interaction === "full" ? "board_direct" : "command_panel", ...details};
    try {
      applyEvent(item);
    } catch (error) {
      model.helpers.setReadout(`INVALID · ${error.message.toUpperCase()}`, "error");
      return;
    }
    model.events.push(item);
    model.selected = model.units[item.unit_id]?.cells.length ? item.unit_id : model.selected;
    update();
  }

  function choose(unitId) {
    const unit = model.units[unitId];
    if (!unit || Number(unit.owner) !== 0 || !unit.cells.length) return;
    model.selected = unitId;
    model.helpers.setReadout(`${unit.label} SELECTED · READ THE HIGHLIGHTED SEAMS`, "idle");
    update();
  }

  function cellAction(row, col) {
    const unitId = Object.keys(model.units).find((id) => cellsKey(model.units[id].cells).has(`${row},${col}`));
    if (unitId && Number(model.units[unitId].owner) === 0) {
      if (model.interaction === "full" && model.selected === unitId) {
        for (const direction of Object.keys(DIRS)) {
          const next = validTranslation(model.selected, direction);
          if (next && cellsKey(next).has(`${row},${col}`)) {
            event("move", {unit_id: model.selected, direction});
            return;
          }
        }
      }
      choose(unitId);
      return;
    }
    if (!model.selected || model.interaction !== "full") return;
    const selected = model.units[model.selected];
    if (!selected || !selected.cells.length) return;
    for (const direction of Object.keys(DIRS)) {
      const next = validTranslation(model.selected, direction);
      if (next && cellsKey(next).has(`${row},${col}`)) {
        event("move", {unit_id: model.selected, direction});
        return;
      }
    }
    if (unitId && Number(model.units[unitId].owner) === 1 && adjacent(selected.cells, model.units[unitId].cells)) {
      event("attack", {unit_id: model.selected, target_id: unitId});
    } else {
      model.helpers.setReadout("CHOOSE A GOLD MOVE SQUARE OR NEIGHBORING RIVAL", "error");
    }
  }

  function boardMarkup() {
    const byCell = {};
    Object.values(model.units).forEach((unit) => unit.cells.forEach(([row, col]) => { byCell[`${row},${col}`] = unit; }));
    const cells = [];
    for (let row = 0; row < model.world.height; row += 1) {
      for (let col = 0; col < model.world.width; col += 1) {
        const unit = byCell[`${row},${col}`];
        const selected = unit && unit.id === model.selected;
        let legal = false;
        if (model.selected && model.interaction === "full") {
          const selectedUnit = model.units[model.selected];
          legal = Object.keys(DIRS).some((direction) => validTranslation(model.selected, direction)?.some(([r, c]) => r === row && c === col));
          if (unit && Number(unit.owner) === 1 && selectedUnit && adjacent(selectedUnit.cells, unit.cells)) legal = true;
        }
        const seam = model.world.seams[row] || {offset: 0, tone: "gold"};
        const classes = ["patchwork-cell", unit ? `is-${Number(unit.owner) === 0 ? "player" : "rival"}` : "is-empty", selected ? "is-selected" : "", legal ? "is-legal" : ""].filter(Boolean).join(" ");
        cells.push(`<button type="button" class="${classes}" data-cell="${row},${col}" data-unit-id="${unit ? esc(unit.id) : ""}" aria-label="${unit ? esc(unit.label) : "empty quilt square"} at row ${row + 1}, column ${col + 1}" style="--seam:${(Number(seam.offset) + col) % 4};--tone:${esc(seam.tone)}"><span>${unit ? esc(unit.id) : ""}</span></button>`);
      }
    }
    return cells.join("");
  }

  function panelMarkup() {
    const selected = model.selected ? model.units[model.selected] : null;
    const playerRoster = Object.values(model.units).filter((unit) => Number(unit.owner) === 0).map((unit) => `<button type="button" class="patchwork-roster-unit ${unit.id === model.selected ? "is-selected" : ""} ${unit.cells.length ? "" : "is-gone"}" data-select-unit="${esc(unit.id)}"><span class="roster-swatch palette-${esc(unit.palette)}"></span><b>${esc(unit.label)}</b><small>${unit.cells.length ? `${unit.cells.length} CELLS · ${unit.moves_left} MOVES` : "REMOVED"}</small></button>`).join("");
    const rivals = Object.values(model.units).filter((unit) => Number(unit.owner) === 1 && unit.cells.length).map((unit) => `<button type="button" class="patchwork-attack-target" data-action="attack" data-target="${esc(unit.id)}" ${!selected || !adjacent(selected.cells, unit.cells) || Number(selected.attacks_left) <= 0 ? "disabled" : ""}>STRIKE ${esc(unit.id)} · ${unit.cells.length} CELLS</button>`).join("") || `<span class="patchwork-muted">NO RIVALS IN RANGE</span>`;
    const moveButtons = Object.keys(DIRS).map((direction) => `<button type="button" class="patchwork-move" data-action="move" data-direction="${direction}" ${!selected || !validTranslation(model.selected, direction) || Number(selected.moves_left) <= 0 ? "disabled" : ""}>${direction.toUpperCase()}</button>`).join("");
    return `<div class="patchwork-rail-head"><span>FRIENDLY PATCHES</span><b>${live(0).length} ALIVE</b></div><div class="patchwork-roster">${playerRoster}</div><div class="patchwork-inspector"><span>SELECTED READOUT</span><strong>${selected ? esc(selected.id) : "—"}</strong><small>${selected ? `${selected.cells.length} OCCUPIED CELLS = HEALTH · ${selected.moves_left} MOVES · ${selected.attacks_left} STRIKE` : "SELECT ON THE QUILT"}</small></div>${model.interaction === "simplified" ? `<div class="patchwork-command"><span>COMMAND PANEL</span><div class="patchwork-moves">${moveButtons}</div><div class="patchwork-rivals">${rivals}</div></div>` : `<div class="patchwork-direct-note"><span>BOARD DIRECT</span><p>Select a friendly patch, then click a gold square to translate its whole footprint. Click a neighboring rival to remove one edge cell.</p></div>`}<div class="patchwork-rail-divider"></div><div class="patchwork-threat"><span>RIVAL PATCHES</span><b>${live(1).length} REMAIN</b><small>${model.lastResponse}</small></div>`;
  }

  function update() {
    const root = document.querySelector(".patchwork-skirmish");
    if (!root) return;
    const board = root.querySelector(".patchwork-board");
    if (board) board.innerHTML = boardMarkup();
    const rail = root.querySelector(".patchwork-rail");
    if (rail) rail.innerHTML = panelMarkup();
    root.querySelector("#patchwork-turn").textContent = `TURN ${model.turn + 1} / ${model.world.max_turns}`;
    root.querySelector("#patchwork-event-count").textContent = `${model.events.length} COMMITTED ACTIONS`;
    root.querySelector("#patchwork-status-line").textContent = model.terminal ? (model.won ? "ALL RIVALS REMOVED · CERTIFY THE SURVIVING QUILT" : "YOUR PATCHWORK IS LOST · CERTIFY TO REQUEST A FRESH PATCH") : model.interaction === "full" ? "SELECT A FRIENDLY FOOTPRINT" : "SELECT A PATCH FROM THE ROSTER";
    root.querySelector("#patchwork-end-turn").disabled = model.terminal;
    root.querySelectorAll(".patchwork-cell").forEach((button) => button.addEventListener("click", () => {
      const [row, col] = button.dataset.cell.split(",").map(Number);
      cellAction(row, col);
    }));
    root.querySelectorAll("[data-select-unit]").forEach((button) => button.addEventListener("click", () => choose(button.dataset.selectUnit)));
    root.querySelectorAll("[data-action='move'],[data-action='attack']").forEach((button) => button.addEventListener("click", () => {
      if (button.dataset.target) event("attack", {unit_id: model.selected, target_id: button.dataset.target});
      else event("move", {unit_id: model.selected, direction: button.dataset.direction});
    }));
    root.querySelector("#patchwork-end-turn").onclick = () => event("end_turn", {});
    root.querySelector("#patchwork-submit").onclick = submit;
  }

  async function submit() {
    if (!model || model.submitting) return;
    model.submitting = true;
    const current = model;
    try {
      const response = await fetch("/result", {method: "POST", headers: {"content-type": "application/json"}, body: JSON.stringify({
        mechanic_id: current.state.mechanic_id, task_id: current.state.task_id, challenge_id: current.state.challenge_id, events: current.events, final_state: snapshot(),
      })});
      const outcome = await response.json();
      if (outcome.passed === true) {
        current.completed = true;
        current.helpers.setReadout("PASS", "passed");
        document.querySelector(".patchwork-skirmish")?.setAttribute("data-verdict", "pass");
        return;
      }
      if (outcome.passed === false && outcome.state) {
        await render(outcome.state, current.helpers, {freshFailure: true});
        current.helpers.setReadout("FAIL · FRESH PATCH ISSUED", "error");
      } else {
        current.submitting = false;
        current.helpers.setReadout("RUN REJECTED", "error");
      }
    } catch (_error) {
      if (model === current) {
        current.submitting = false;
        current.helpers.setReadout("VERIFIER LINK LOST · RETRY CERTIFY", "error");
      }
    }
  }

  async function render(state, helpers, options = {}) {
    document.body.dataset.mechanic = ID;
    model = {
      state, helpers, world: state.world, interaction: state.control_condition?.interaction || "full", units: {}, selected: null,
      events: [], turn: 0, terminal: false, won: false, actionCount: 0, enemyActions: 0, lastResponse: "RIVALS WAIT IN THEIR STITCHES", submitting: false, completed: false,
    };
    (state.world.units || []).forEach((raw) => { model.units[raw.id] = {...raw, cells: raw.cells.map((cell) => [...cell]), moves_left: Number(state.world.movement_points), attacks_left: 1}; });
    const root = helpers.app;
    root.innerHTML = `<section class="patchwork-skirmish" data-interaction="${esc(model.interaction)}" data-challenge-id="${esc(state.challenge_id)}" data-fresh-failure="${options.freshFailure ? "true" : "false"}" data-verdict=""><header class="patchwork-header"><div><span class="patchwork-kicker">QUARANTINE QUILT / TACTICAL RECOVERY GRID</span><h1>PATCHWORK SKIRMISH</h1><p>${esc(state.prompt)}</p></div><div class="patchwork-turn"><span id="patchwork-turn">TURN 1 / ${state.world.max_turns}</span><strong>OCCUPIED AREA = HEALTH</strong></div></header><main class="patchwork-main"><section class="patchwork-stage"><div class="patchwork-board" style="--board-cols:${state.world.width};--board-rows:${state.world.height}">${boardMarkup()}</div><div class="patchwork-legend"><span><i class="legend-player"></i>YOUR PATCHES</span><span><i class="legend-rival"></i>RIVAL PATCHES</span><span><i class="legend-legal"></i>LEGAL BOARD INPUT</span></div></section><aside class="patchwork-rail">${panelMarkup()}</aside></main><footer class="patchwork-footer"><div><span id="patchwork-status-line">${model.interaction === "full" ? "SELECT A FRIENDLY FOOTPRINT" : "SELECT A PATCH FROM THE ROSTER"}</span><small id="patchwork-event-count">0 COMMITTED ACTIONS</small><div class="readout" data-status="idle">BOARD LIVE</div></div><div class="patchwork-footer-actions"><button type="button" id="patchwork-end-turn">END TURN ↗</button><button type="button" id="patchwork-submit">${esc(state.submit_label || "CERTIFY SKIRMISH")}</button></div></footer>${helpers.cheatPanelTemplate ? helpers.cheatPanelTemplate() : ""}</section>`;
    window.patchworkSkirmishModel = {snapshot, state: () => model.state, events: () => model.events};
    update();
    if (helpers.installCheatPanel) helpers.installCheatPanel();
    if (options.freshFailure) setTimeout(() => document.querySelector(".patchwork-skirmish")?.setAttribute("data-fresh-failure", "false"), 1100);
  }

  window.WeirdCaptchaMechanics = window.WeirdCaptchaMechanics || {};
  window.WeirdCaptchaMechanics[ID] = {rootSelector: ".patchwork-skirmish", render};
})();

File: weird_captcha_gym/shared_runtime/server/incubator_graders/patchwork_skirmish.py
from __future__ import annotations

import copy
import heapq
import itertools
from typing import Any


MECHANIC_ID = "patchwork_skirmish"
PLAYER = 0
ENEMY = 1
DIRS = {
    "north": (-1, 0),
    "south": (1, 0),
    "west": (0, -1),
    "east": (0, 1),
}


def _live(units: dict[str, dict[str, Any]], owner: int | None = None) -> list[dict[str, Any]]:
    return [
        unit for unit in units.values()
        if unit["cells"] and (owner is None or int(unit["owner"]) == owner)
    ]


def _distance(cells_a: list[list[int]], cells_b: list[list[int]]) -> int:
    return min(abs(a[0] - b[0]) + abs(a[1] - b[1]) for a in cells_a for b in cells_b)


def _adjacent(cells_a: list[list[int]], cells_b: list[list[int]]) -> bool:
    return _distance(cells_a, cells_b) == 1


def _occupied(units: dict[str, dict[str, Any]], skip: str | None = None) -> set[tuple[int, int]]:
    return {
        (int(row), int(col))
        for unit_id, unit in units.items()
        if unit_id != skip and unit["cells"]
        for row, col in unit["cells"]
    }


def _translated(cells: list[list[int]], direction: str) -> list[list[int]]:
    dr, dc = DIRS[direction]
    return [[int(row) + dr, int(col) + dc] for row, col in cells]


def _valid_translation(
    units: dict[str, dict[str, Any]], unit_id: str, direction: str, width: int, height: int
) -> list[list[int]] | None:
    if direction not in DIRS:
        return None
    cells = _translated(units[unit_id]["cells"], direction)
    if any(row < 0 or row >= height or col < 0 or col >= width for row, col in cells):
        return None
    occupied = _occupied(units, skip=unit_id)
    if any((row, col) in occupied for row, col in cells):
        return None
    return cells


def _remove_edge_cell(unit: dict[str, Any], power: int) -> int:
    removed = min(max(0, int(power)), len(unit["cells"]))
    if removed:
        del unit["cells"][-removed:]
    return removed


def initial_state(world: dict[str, Any]) -> dict[str, Any]:
    units = {}
    for raw in world.get("units") or []:
        unit = copy.deepcopy(raw)
        unit["cells"] = [[int(cell[0]), int(cell[1])] for cell in unit["cells"]]
        unit["moves_left"] = int(world["movement_points"])
        unit["attacks_left"] = 1
        units[str(unit["id"])] = unit
    return {
        "turn": 0,
        "units": units,
        "terminal": False,
        "won": False,
        "action_count": 0,
        "enemy_actions": 0,
        "last_response": "RIVALS WAIT IN THEIR STITCHES",
    }


def _nearest_enemy(units: dict[str, dict[str, Any]], unit: dict[str, Any]) -> dict[str, Any] | None:
    candidates = _live(units, ENEMY if int(unit["owner"]) == PLAYER else PLAYER)
    if not candidates:
        return None
    return min(candidates, key=lambda other: (_distance(unit["cells"], other["cells"]), str(other["id"])))


def _enemy_response(state: dict[str, Any], world: dict[str, Any]) -> None:
    notes: list[str] = []
    width, height = int(world["width"]), int(world["height"])
    for enemy in sorted(_live(state["units"], ENEMY), key=lambda item: str(item["id"])):
        if not enemy["cells"]:
            continue
        target = _nearest_enemy(state["units"], enemy)
        if target is None:
            break
        if _adjacent(enemy["cells"], target["cells"]):
            removed = _remove_edge_cell(target, 1)
            state["enemy_actions"] += 1
            notes.append(f"{enemy['id']} chips {target['id']} ({removed} cell)")
            continue
        steps = int(world["enemy_movement_points"])
        for _ in range(steps):
            target = _nearest_enemy(state["units"], enemy)
            if target is None or not enemy["cells"]:
                break
            choices = []
            for direction in DIRS:
                cells = _valid_translation(state["units"], str(enemy["id"]), direction, width, height)
                if cells is not None:
                    choices.append((_distance(cells, target["cells"]), direction, cells))
            if not choices:
                break
            choices.sort(key=lambda item: (item[0], ("north", "west", "south", "east").index(item[1])))
            _, direction, cells = choices[0]
            enemy["cells"] = cells
            state["enemy_actions"] += 1
            notes.append(f"{enemy['id']} drifts {direction}")
    state["last_response"] = "; ".join(notes) if notes else "RIVALS HAVE NO LEGAL RESPONSE"


def apply_event(state: dict[str, Any], world: dict[str, Any], event: dict[str, Any]) -> None:
    if state["terminal"]:
        raise ValueError("the skirmish is already over")
    event_type = str(event.get("type") or "")
    units = state["units"]
    if event_type == "move":
        unit_id = str(event.get("unit_id") or "")
        direction = str(event.get("direction") or "")
        unit = units.get(unit_id)
        if unit is None or int(unit["owner"]) != PLAYER or not unit["cells"]:
            raise ValueError("move must select a living player patch")
        if int(unit["moves_left"]) <= 0:
            raise ValueError("that patch has spent its movement")
        cells = _valid_translation(units, unit_id, direction, int(world["width"]), int(world["height"]))
        if cells is None:
            raise ValueError("move collides with the board or another patch")
        unit["cells"] = cells
        unit["moves_left"] -= 1
    elif event_type == "attack":
        unit_id = str(event.get("unit_id") or "")
        target_id = str(event.get("target_id") or "")
        unit = units.get(unit_id)
        target = units.get(target_id)
        if unit is None or target is None or int(unit["owner"]) != PLAYER or int(target["owner"]) != ENEMY:
            raise ValueError("attack must select a player and rival patch")
        if not unit["cells"] or not target["cells"] or int(unit["attacks_left"]) <= 0:
            raise ValueError("that patch cannot attack now")
        if not _adjacent(unit["cells"], target["cells"]):
            raise ValueError("the rival patch is not neighboring")
        _remove_edge_cell(target, int(world["attack_power"]))
        unit["attacks_left"] -= 1
    elif event_type == "end_turn":
        if not _live(units, ENEMY):
            state["terminal"] = True
            state["won"] = bool(_live(units, PLAYER))
            return
        _enemy_response(state, world)
        state["turn"] += 1
        for unit in _live(units):
            if int(unit["owner"]) == PLAYER:
                unit["moves_left"] = int(world["movement_points"])
                unit["attacks_left"] = 1
        if not _live(units, PLAYER) or state["turn"] >= int(world["max_turns"]):
            state["terminal"] = True
            state["won"] = False
    else:
        raise ValueError("unknown patchwork action")
    state["action_count"] += 1


def summary(state: dict[str, Any]) -> dict[str, Any]:
    return {
        "turn": int(state["turn"]),
        "units": {
            unit_id: {
                "cells": copy.deepcopy(unit["cells"]),
                "moves_left": int(unit["moves_left"]),
                "attacks_left": int(unit["attacks_left"]),
            }
            for unit_id, unit in sorted(state["units"].items())
        },
        "terminal": bool(state["terminal"]),
        "won": bool(state["won"]),
        "action_count": int(state["action_count"]),
        "enemy_actions": int(state["enemy_actions"]),
        "last_response": str(state["last_response"]),
    }


def _search_key(state: dict[str, Any]) -> tuple[Any, ...]:
    """Return the future-relevant portion of a replay state."""
    return (
        int(state["turn"]),
        bool(state["terminal"]),
        bool(state["won"]),
        tuple(
            (
                unit_id,
                tuple(tuple(cell) for cell in unit["cells"]),
                int(unit["moves_left"]),
                int(unit["attacks_left"]),
            )
            for unit_id, unit in sorted(state["units"].items())
        ),
    )


def _search_score(state: dict[str, Any], depth: int) -> float:
    """Prefer damage, adjacency, and surviving area during oracle search."""
    if state["terminal"]:
        return -100000.0 if state["won"] else 100000.0
    players = _live(state["units"], PLAYER)
    enemies = _live(state["units"], ENEMY)
    enemy_cells = sum(len(unit["cells"]) for unit in enemies)
    player_cells = sum(len(unit["cells"]) for unit in players)
    distances = (
        sum(_distance(player["cells"], enemy["cells"]) for player in players for enemy in enemies)
        if players and enemies
        else 0
    )
    adjacent_pairs = sum(
        1 for player in players for enemy in enemies if _adjacent(player["cells"], enemy["cells"])
    )
    fragile = sum(max(0, 2 - len(unit["cells"])) for unit in players)
    return enemy_cells * 30 - player_cells * 2 + distances * 0.5 - adjacent_pairs * 20 + fragile * 40 + depth * 0.01


def _search_actions(state: dict[str, Any], world: dict[str, Any]) -> list[dict[str, Any]]:
    actions: list[dict[str, Any]] = []
    for unit_id, unit in sorted(state["units"].items()):
        if int(unit["owner"]) != PLAYER or not unit["cells"]:
            continue
        if int(unit["attacks_left"]) > 0:
            for target_id, target in sorted(state["units"].items()):
                if int(target["owner"]) == ENEMY and target["cells"] and _adjacent(unit["cells"], target["cells"]):
                    actions.append({"type": "attack", "input_source": "board_direct", "unit_id": unit_id, "target_id": target_id})
        if int(unit["moves_left"]) > 0:
            for direction in ("north", "west", "south", "east"):
                if _valid_translation(state["units"], unit_id, direction, int(world["width"]), int(world["height"])) is not None:
                    actions.append({"type": "move", "input_source": "board_direct", "unit_id": unit_id, "direction": direction})
    actions.append({"type": "end_turn", "input_source": "turn_button"})
    return actions


def _search_plan(world: dict[str, Any]) -> list[dict[str, Any]] | None:
    """Find a bounded winning replay when the greedy oracle is unsafe."""
    initial = initial_state(world)
    counter = itertools.count()
    queue: list[tuple[float, int, dict[str, Any], list[dict[str, Any]]]] = [
        (_search_score(initial, 0), next(counter), initial, [])
    ]
    best_depth = {_search_key(initial): 0}
    expanded = 0
    max_depth = int(world["max_turns"]) * 5 + 20
    while queue and expanded < 20000:
        _, _, state, path = heapq.heappop(queue)
        expanded += 1
        if state["terminal"]:
            if state["won"]:
                return path
            continue
        if len(path) >= max_depth:
            continue
        for action in _search_actions(state, world):
            next_state = copy.deepcopy(state)
            try:
                apply_event(next_state, world, action)
            except (KeyError, TypeError, ValueError):
                continue
            depth = len(path) + 1
            key = _search_key(next_state)
            if key in best_depth and best_depth[key] <= depth:
                continue
            best_depth[key] = depth
            next_event = {"sequence": depth, **action}
            heapq.heappush(
                queue,
                (_search_score(next_state, depth), next(counter), next_state, path + [next_event]),
            )
    return None


def _failure(feedback: str) -> dict[str, Any]:
    return {"graded": True, "passed": False, "score": 0, "feedback": feedback}


def grade(payload: dict[str, Any], ground_truth: dict[str, Any], public_state: dict[str, Any]) -> dict[str, Any]:
    if payload.get("mechanic_id") != MECHANIC_ID or ground_truth.get("mechanic_id") != MECHANIC_ID or public_state.get("mechanic_id") != MECHANIC_ID:
        return _failure("mechanic mismatch")
    challenge_id = str(ground_truth.get("challenge_id") or "")
    task_id = str(ground_truth.get("task_id") or "")
    if payload.get("challenge_id") != challenge_id or public_state.get("challenge_id") != challenge_id:
        return _failure("stale challenge")
    if payload.get("task_id") != task_id or public_state.get("task_id") != task_id:
        return _failure("task identity mismatch")
    if public_state.get("world") != ground_truth.get("world"):
        return _failure("public world does not match replay world")
    if public_state.get("control_condition") != ground_truth.get("control_condition"):
        return _failure("public condition does not match replay condition")
    world = ground_truth.get("world")
    if not isinstance(world, dict):
        return _failure("missing world")
    interaction = str((ground_truth.get("control_condition") or {}).get("interaction") or "full")
    expected_source = "board_direct" if interaction == "full" else "command_panel"
    events = payload.get("events")
    if not isinstance(events, list) or not events or len(events) > 500:
        return _failure("missing or oversized action transcript")
    state = initial_state(world)
    for index, event in enumerate(events, start=1):
        if not isinstance(event, dict) or int(event.get("sequence", -1)) != index:
            return _failure(f"action {index} has an invalid sequence")
        event_type = str(event.get("type") or "")
        expected_event_source = "turn_button" if event_type == "end_turn" else expected_source
        if event.get("input_source") != expected_event_source:
            return _failure(f"action {index} used the wrong interaction surface")
        try:
            apply_event(state, world, event)
        except (KeyError, TypeError, ValueError) as exc:
            return _failure(f"action {index} is invalid: {exc}")
    expected = summary(state)
    if payload.get("final_state") != expected:
        return _failure("submitted final state does not match deterministic replay")
    passed = bool(state["terminal"] and state["won"] and _live(state["units"], PLAYER) and not _live(state["units"], ENEMY))
    return {
        "graded": True,
        "passed": passed,
        "score": 100 if passed else 0,
        "feedback": f"turn {state['turn']}; player patches {len(_live(state['units'], PLAYER))}; rival patches {len(_live(state['units'], ENEMY))}; enemy responses {state['enemy_actions']}",
    }


def plan(world: dict[str, Any]) -> list[dict[str, Any]]:
    """Find a robust ordinary-input plan for evidence and unit tests."""
    state = initial_state(world)
    events: list[dict[str, Any]] = []
    source = "board_direct"

    def issue(event_type: str, **details: Any) -> None:
        event_source = "turn_button" if event_type == "end_turn" else source
        event = {"sequence": len(events) + 1, "type": event_type, "input_source": event_source, **details}
        apply_event(state, world, event)
        events.append(event)

    # A bounded greedy controller uses the same visible cells that a human
    # reads. It closes on the nearest rival, chips it, then reuses the patch.
    for _ in range(int(world["max_turns"]) + 2):
        if state["terminal"]:
            break
        for unit in sorted(_live(state["units"], PLAYER), key=lambda item: str(item["id"])):
            unit_id = str(unit["id"])
            if not unit["cells"]:
                continue
            while int(unit["moves_left"]) > 0 and _live(state["units"], ENEMY):
                target = _nearest_enemy(state["units"], unit)
                if target is None or _adjacent(unit["cells"], target["cells"]):
                    break
                choices = []
                for direction in DIRS:
                    cells = _valid_translation(state["units"], unit_id, direction, int(world["width"]), int(world["height"]))
                    if cells is not None:
                        choices.append((_distance(cells, target["cells"]), direction))
                if not choices:
                    break
                choices.sort(key=lambda item: (item[0], ("north", "west", "south", "east").index(item[1])))
                issue("move", unit_id=unit_id, direction=choices[0][1])
            target = _nearest_enemy(state["units"], unit)
            if target is not None and int(unit["attacks_left"]) > 0 and _adjacent(unit["cells"], target["cells"]):
                issue("attack", unit_id=unit_id, target_id=str(target["id"]))
        issue("end_turn")
        if state["terminal"]:
            break
    if state["terminal"] and state["won"]:
        return events
    searched = _search_plan(world)
    if searched is None:
        raise ValueError("patchwork oracle could not find a winning plan")
    return searched


def cheat(public_state: dict[str, Any], ground_truth: dict[str, Any]) -> dict[str, Any]:
    return {
        "instruction": "Select a patch, translate its whole footprint toward a neighboring rival, chip one edge cell, and end each turn to read the response.",
        "plan_length": len(plan(ground_truth["world"])),
    }
